### PR TITLE
DOSE-705 zpool iostat histograms

### PR DESCRIPTION
**Pull request by don-brady**
### Motivation and Context
This is a follow on to DOSE-475 (#494) where object store iostats were added.  This PR adds histograms for latency and request size as well as active counts (count of inflight gets/puts).

### Description
The Object Agent adds these new histograms and active counter to the existing object stats payload.
There was some minor refactoring to previous object store stat code to accommodate these new stats.

**Note:**  the first commit to this PR does not have the man page updates or the usage updates for these new options.
### CLI Overview

All the object store based iostats make use of the `'-o'` argument. In the following examples, the pool name is `observe`.

Observing latencies `[-ol]` (now with real values).
```
delphix@ip-10-110-215-29:~$ zpool iostat -ol observe 1
             --------- object_agent_total ---------    --------- object_store_total ---------
             operations    throughput      latency     operations    throughput      latency
pool         read  write   read  write   read  write    get    put    get    put    get    put
----------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
observe      3.24     27   407K  1.40M   78ms  132ms   1.24   4.39   738K  2.00M  125ms  171ms
observe       665      0  83.1M      0   71ms   85ns    127   1.00   127M    188  137ms    4ms
observe       462      0  57.7M      0   85ms   85ns    108   1.00   108M    188  164ms    9ms
observe       232      0  29.0M      0   86ms   85ns     29   1.00  29.0M    188  143ms   36ms
observe       848      0   106M      0   62ms   85ns    107   1.00   107M    188  130ms    9ms
observe       744     13  92.9M  20.5K   62ms   25ms     95   8.99  94.0M  26.3K  127ms   39ms
observe       927      0   116M      0   62ms   85ns    116   1.00   116M    188  125ms    4ms
observe       752      0  93.9M      0   64ms   85ns     95   1.00  94.5M    188  129ms    4ms
observe       817      0   102M      0   61ms   85ns    104   1.00   104M    188  124ms    4ms
observe       785      0  98.0M      0   62ms   85ns    125   1.00   124M    188  122ms    9ms
observe       516    113  64.3M  6.56M   63ms  102ms    119     16   118M  6.57M  119ms   87ms
observe       706      0  88.1M      0   60ms   85ns    153   1.00   153M    188  118ms    9ms
observe       458      0  56.8M      0   72ms   85ns     65   1.00  64.9M    188  140ms    4ms
observe       696      0  86.9M      0   74ms   85ns     88   1.00  87.9M    188  136ms    9ms
observe       664      0  82.9M      0   69ms   85ns     84   1.00  83.8M    188  140ms    9m
^C
```

Observing active counts `[-oq]` (for each of the data/metadata/reclaim categories).
```
              --- object_agent_io ---    ---------- object_store_data ---------    -------- object_store_metadata -------    -------- object_store_reclaim --------
             operations    throughput    operations    throughput      active      operations    throughput      active      operations    throughput      active
pool         read  write   read  write    get    put    get    put    get    put    get    put    get    put    get    put    get    put    get    put    get    put
----------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
observe         0     90      0   335K      0   3.77      0   317K      0      0      0   4.71      0  11.0M      0      1      0      0      0      0      0      0
observe         0  12.2K      0  49.1M      0     54      0  54.3M      0     18     23   7.99  9.36M  7.95M      0      0   7.99   7.99  8.07M    479      0      0
observe         0  26.6K      0  93.4M      0     93      0  93.3M      0     13      0   0.98      0    183      0      0      0      0      0      0      0      0
observe         0  32.1K      0   118M      0    117      0   118M      0      2      0   0.98      0    185      0      0      0      0      0      0      0      0
observe         0  31.6K      0   126M      0    145      0   146M      0      1      0   1.00      0    188      0      0      0      0      0      0      0      0
observe         0  38.7K      0   146M      0    129      0   130M      0      1      0   0.86      0    161      0      0      0      0      0      0      0      0
observe         0  30.6K      0   114M      0    148      0   149M      0      1      0   1.00      0    188      0      0      0      0      0      0      0      0
observe         0  17.3K      0  51.8M      0     26      0  22.6M      0      0   7.69     10  16.8M  21.2M      2      0      0      0      0      0      0      0
observe         0  13.3K      0  55.3M      0     67      0  67.3M      0     17   2.00   5.99  64.3K   937K      5      0      0      0      0      0      0      0
observe         0  38.6K      0   128M      0    125      0   126M      0     16      0   1.78      0    335      0      0      0      0      0      0      0      0
observe         0  38.3K      0   136M      0    133      0   134M      0      0      0   0.95      0    179      0      0      0      0      0      0      0      0
observe         0  22.2K      0  81.7M      0     80      0  77.1M      0      0   8.73   4.85  21.2M   659K      1      1      0      0      0      0      0      0
observe         0  11.9K      0  21.5M      0     22      0  22.2M      0      6   2.99   1.99  3.74M    293      5      0     16   8.98  15.9M  1.96M      0      0
observe         0  59.0K      0  31.6M      0     33      0  32.8M      0      0      0   0.96      0    181      0      0      0      0      0      0      0      0
observe         0  1.36K      0   697K      0   2.95      0  5.66K      0      0   2.95     10  2.88M  12.1M      0      0      0      0      0      0      0      0
observe         0  7.84K      0  26.5M      0     27      0  26.9M      0     10      0   0.99      0    186      0      0      0      0      0      0      0      0
observe         0  20.0K      0  68.5M      0     70      0  71.0M      0      3      0   0.99      0    187      0      0      0      0      0      0      0      0
observe         0  36.6K      0  30.6M      0     30      0  27.4M      0      0      0   0.91      0    171      0      0      0      0      0      0      0      0
observe         0      0      0      0      0      0      0      0      0      0   6.99   5.99  9.95M  1.41M      0      0     59     29  56.9M  3.55M      0      0
observe         0      0      0      0      0      0      0      0      0      3      0   1.00      0    188      0      0      0      0      0      0      0      0
observe         0  31.2K      0  17.4M      0     26      0  26.6M      0      0      0   0.99      0    187      0      0      0      0      0      0      0      0
observe         0  30.8K      0  16.5M      0     13      0  10.9M      0      0   2.09   6.26   998K  4.81M      0      0      0      0      0      0      0      0
observe         0      0      0      0      0      0      0      0      0      0      0   1.00      0    188      0      0      0      0      0      0      0      0
observe         0      0      0      0      0      0      0      0      0      0      0   1.00      0    188      0      0      0      0      0      0      0      0
observe         0  20.5K      0  40.9M      0     66      0  66.3M      0     14      0   1.00      0    187      0      0      0      0      0      0      0      0
observe         0  41.8K      0  76.9M      0     61      0  58.5M      0      0      0   2.27      0  5.52K      1      2      0      0      0      0      0      0
observe         0      0      0      0      0      0      0      0      0      0   1.00   4.99  9.30K  18.4K      5      0      0      0      0      0      0      0
observe         0      0      0      0      0      0      0      0      0      0      0   1.00      0    188      0      0      0      0      0      0      0      0
^C
```

Observing request size histogram `[-or]`
```
delphix@ip-10-110-215-29:~$ zpool iostat -or observe

observe      synchronous  asynchronous  object_store
req_size     read  write   read  write    get    put
----------  -----  -----  -----  -----  -----  -----
64              0      0      0      0      0    548
128             0      0      0      0      2    268
256             0      0      0      0     10  1.18K
512           730      0  15.7K  5.82M     40     64
1K          2.05K      0  18.9K   561K     81    337
2K            582      0   134K  5.17M     46     63
4K            321      0   104K  3.31M    106    651
8K          1.20K      0  12.4K   280K    228    401
16K           411      0    601  7.04K    111    196
32K           577      0    739  9.12K     74    101
64K             0      0      0      0    102    136
128K            0      0      0      0     67     98
256K            0      0      0      0     94    177
512K            0      0      0      0    222    707
1M              0      0      0      0    758  2.74K
2M              0      0      0      0  9.32K  36.9K
4M              0      0      0      0    316    318
8M              0      0      0      0      0      0
16M             0      0      0      0      0      0
----------------------------------------------------
```

Observing latency histogram `[-ow]`
```
delphix@ip-10-110-215-29:~$ zpool iostat -ow observe

observe     object_agent      object_store
latency      read  write    get    put    del
----------  -----  -----  -----  -----  -----
1ns             0      0      0      0      0
3ns             0      0      0      0      0
7ns             0      0      0      0      0
15ns            0      0      0      0      0
31ns            0      0      0      0      0
63ns            0      0      0      0      0
127ns           0      0      0      0      0
255ns           0      0      0      0      0
511ns           0      0      0      0      0
1us             0      0      0      0      0
2us             0      0      0      0      0
4us             0      0      0      0      0
8us             0      0      0      0      0
16us            0      0      0      0      0
32us            0      0      0      0      0
65us            0      0      0      0      0
131us         939      0      0      0      0
262us       9.66K      0      0      0      0
524us       10.4K      0      0      0      0
1ms         3.40K      0      0      0      0
2ms         3.34K      0      0      0      0
4ms         3.92K      0      0      0      0
8ms         3.49K      0      0      0      0
16ms        3.88K      0      0      0      0
33ms        10.2K    669     72     35      0
67ms        98.1K  86.9K    786  2.14K    134
134ms        132K  2.38M  1.88K  4.76K      8
268ms       11.4K  7.14M  4.89K  29.3K      4
536ms         784  5.14M  2.86K  9.85K      9
1s            246  1021K  1.08K    495      8
2s              0  1.11K     69     22     51
4s              0    702      2      1      3
8s              0    459      0      0      0
17s             0    161      0      0      0
34s             0      0      0      0      0
68s             0      0      0      0      0
137s            0      0      0      0      0
---------------------------------------------
```

### How Has This Been Tested?
Manual testing of all the variants.  Confirmed that legacy iostat variants still work.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### CODE REVIEW

**Comment by ahrens** [comment]
In the example output of `zpool iostat -ol observe 1`, how come we have some seconds where `object_store_total : operations : get` is nonzero but `object_store_total : latency : get` is `-`?  Shouldn't we be able to calculate the average latency of those operations?

Clippy has some useful suggestions for you, listed in the checkstyle output.

File: cmd/zpool/zpool_main.c [hunk 1]
```diff
@@ -5007,14 +5125,20 @@ print_vdev_stats(zpool_handle_t *zhp, const char *name, nvlist_t *oldnv,
 			scale = (double)NANOSEC / tdelta;
 	}
 
-	if (cb->cb_flags & IOS_DEFAULT_M) {
+	if ((cb->cb_flags & IOS_ANYHISTO_M) == 0)
```
Comment (tonynguien) [inline cmd/zpool/zpool_main.c:5128]: Should this be `if ((cb->cb_flags & (IOS_ANYHISTO_M | IOS_DEFAULT_M))` so we calculate calcvs in the case where cb->cb_flags has IOS_DEFAULT_M set?

File: cmd/zpool/zpool_main.c [hunk 2]
```diff
@@ -4889,27 +5002,32 @@ print_iostat_object_latency(nvlist_t *oldnv, nvlist_t *newnv,
...
 	nvlist_t *oldoss = NULL, *newoss;
 	newoss = fnvlist_lookup_nvlist(newnv, ZPOOL_CONFIG_OBJECT_STORE_STATS);
-	if (oldnv == NULL ||
-	    nvlist_lookup_nvlist(oldnv, ZPOOL_CONFIG_OBJECT_STORE_STATS,
-	    &oldoss) != 0) {
-		oldoss = fnvlist_alloc();
+
+	if (oldnv != NULL) {
+		(void) nvlist_lookup_nvlist(oldnv,
+		    ZPOOL_CONFIG_OBJECT_STORE_STATS, &oldoss);
```
Comment (tonynguien) [inline cmd/zpool/zpool_main.c:5008]: is there a reason we don't use `oldoss = fnnvlist_lookup_nvlist(oldnv, ZPOOL_CONFIG_OBJECT_STORE_STATS);` for consistency?

**Review by tonynguien — APPROVED** [review]
I have a couple minor comments. It looks good to me.

File: cmd/zpool/zpool_main.c (hunk 1 above)
Comment (don-brady) [reply cmd/zpool/zpool_main.c:5128]: Histograms are mutually exclusive to all other iostatsi, including `IOS_DEFAULT_M`.  Added a comment for clarity:
```

+       /* When not asking for histograms, we need to generate the calcvs */
        if ((cb->cb_flags & IOS_ANYHISTO_M) == 0)
                calc_default_iostats(oldvs, newvs, calcvs);
```

**Pull request merged**